app/services/ipam/external_ipam.rb

```diff
@@ -0,0 +1,32 @@
+module IPAM
+  class ExternalIpam < Base
+    delegate :external_ipam_proxy, :to => :subnet
+
+    def suggest_ip
+      if self.mac.nil?
+        errors.add(:mac, "You must specify a MAC address before selecting the External IPAM subnet")
+        return nil
+      end
+
+      logger.debug "Obtaining next available IP from IPAM for subnet #{@subnet.network_address}"
+      response = external_ipam_proxy.next_ip(@subnet, self.mac)
+
+      if response.key?('error')
+        errors.add(:subnet, response['error'])
+        nil
+      else
+        next_ip = response["data"]
+        logger.debug("IPAM returned #{next_ip} as the next available IP in subnet #{@subnet.network_address}")
+        next_ip
+      end
+    rescue => e
+      logger.warn "Failed to fetch the next available IP address from IPAM: #{e}"
+      errors.add(:subnet, _('Failed to fetch the next available IP address from IPAM %{message}') % {:message => e})
+      nil
+    end
+
+    def suggest_new?
+      false
+    end
+  end
+end
```

Comment:
There is another major problem here which is by our design. Our current IPAMs are either IPv4 or IPv6, there is none which handles both. Users can either set one or another for a subnet. I believe that External IPAM should cover both, therefore I suggest this:

Extend subnet_controller#freeip action to also have 4 or 6 parameter which would be passed into IPAM constructor so the this implementation (the method `suggest_ip`) knowns if it should fetch IPv4 or IPv6. This requires a small change in our UI so the HTML/JS form sends the constant along.

@ekohl does that sound good to you:

Reply:
@lzap Looked into this a little bit and it seems that the `subnet_controller#freeip` action only fires when the IP's are "suggested" on subnet selection. Since we are persisting the IP in External IPAM only when the Host is created(or updated), queuing up separate jobs for IPv4 and IPv6 will be problematic. 

Suppose a user creates a host with a dual stack interface, and the IPv6 address is available, but the IPv4 is taken. Since they are queued in separate jobs, there is no easy way for one job to know if the other one succeeded or failed etc. 

A suggestion might be to create one job with both the IPv4 and IPv6 addresses, and add/remove both IP's in the same job, given we are dealing with an external system. This will give us a bit more control and treat it more like a transaction.

This does go against some design principles, however we are only dealing with 2 different IP types(4 & 6), and likely won't see any others anytime soon. Let me know what you think.

Reply:
The way it works currently is that freeip call is made before the final transactions, therefore the call only "pre-reserves" an IP address, smart-proxy DHCP module automatically "remembers" them as pre-reserved for a MAC address. This is also called from AJAX form *before* Submit button is hit.

Then, only after Submit is clicked the transaction (or orchestration) happens. DHCP module on smart-proxy is asked to do reservation and it clears the address from the pool of "pre-reserved" ones. This time we prevent race conditions when multiple hosts were given a same IP address. Those pre-reservations are kept for several minutes. Unless we change our design radically, I think the best way is to stick with this. I think I've pointed out on the code in our smart-proxy PR where you could reuse the class.

> Suppose a user creates a host with a dual stack interface, and the IPv6 address is available, but the IPv4 is taken. Since they are queued in separate jobs, there is no easy way for one job to know if the other one succeeded or failed etc.

It's not possible that IPv4 is taken, the purpose of freeip call is to find a free IPv4 address or return error where there is none. The contract is "pre-reserve me an address and give it to me". Or I am not following here.

If you design two separate orchestration actions, they should both succeed. If there is a DHCP conflict, we have a special handling and "Overwrite" option is shown to the user to force rewrite of reservation.

>  suggestion might be to create one job with both the IPv4 and IPv6 addresses

I am not against queuing just single action if this is a single HTTP call to the myIPAM. However if this is gonna be two calls, then two actions are probably better.

Reply:
Thanks @lzap - I should have provided more clarification on the specific use case I am referring to. When a user creates a dual stack interface, and both IPv4 and IPv6 addresses are suggested successfully, suppose the user manually changes one of the suggested IP's in the UI, to an IP that is taken. In this case, they would click submit, then an error would be returned(saying IP taken), but the other IP orchestration would be successful. Clicking submit again, after correcting the "taken" IP would result in the other IP being taken, and would confuse the user. 

I know this doesn't make a lot of sense for a user to do this, but it is a possibility nonetheless. 

Should we worry about handling this particular scenario?

Reply:
> but the other IP orchestration would be successful

In that case our orchestration framework does "rollback" by calling "delXXX" method (opposite of "setXXX" method). This is what we call "transaction compensation" and it's the best what we can achieve in the environment without distributed transaction support.

Reply:
Right - I forgot about the "delXXX" method. I will look into that.